File: src/main/java/com/babylonml/backend/training/operations/Constant.java
package com.babylonml.backend.training.operations;

import com.babylonml.backend.cpu.TensorOperations;
import com.babylonml.backend.training.execution.TensorPointer;
import com.babylonml.backend.training.execution.TrainingExecutionContext;

import it.unimi.dsi.fastutil.ints.IntImmutableList;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

import java.util.List;


@SuppressWarnings("unused")
public final class Constant extends AbstractOperation implements StartOperation {
    private final Tensor constant;

    public Constant(@NonNull TrainingExecutionContext executionContext, Tensor constant) {
        this(null, executionContext, constant);
    }

    public Constant(@Nullable String name, @NonNull TrainingExecutionContext executionContext, Tensor constant) {
        super(name, executionContext, null, null);
        this.constant = constant;
    }

    @Override
    public @NonNull IntImmutableList getMaxResultShape() {
        return constant.getShape();
    }

    @Override
    public @NonNull TensorPointer forwardPassCalculation() {
        var result = executionContext.allocateForwardMemory(this, constant.getShape());

        var stride = TensorOperations.stride(constant.getShape());
        System.arraycopy(constant.getData(), 0, result.buffer(), result.offset(), stride);

        return result;
    }

    @Override
    public @NonNull TensorPointer leftBackwardDerivativeChainValue() {
        return TrainingExecutionContext.NULL;
    }

    @Override
    public @NonNull TensorPointer rightBackwardDerivativeChainValue() {
        return TrainingExecutionContext.NULL;
    }

    @Override
    public @NonNull List<IntImmutableList> getForwardMemoryAllocations() {
        return List.of(constant.getShape());
    }

    @Override
    public @NonNull List<IntImmutableList> getBackwardMemoryAllocations() {
        return List.of();
    }

    @Override
    public boolean requiresBackwardDerivativeChainValue() {
        return false;
    }

    @Override
    public void calculateGradientUpdate() {
        // No gradient update required
    }
}

File: src/main/kotlin/com/babylonml/backend/training/operations/Constant.kt
package com.babylonml.backend.training.operations

import com.babylonml.backend.common.CommonTensorOperations
import com.babylonml.backend.training.execution.TensorPointer
import com.babylonml.backend.training.execution.TrainingExecutionContext
import it.unimi.dsi.fastutil.ints.IntImmutableList


@Suppress("unused")
class Constant(
    name: String?,
    executionContext: TrainingExecutionContext,
    private val constant: Tensor
) : AbstractOperation(name, executionContext, null, null), StartOperation {
    constructor(executionContext: TrainingExecutionContext, constant: Tensor) : this(
        null,
        executionContext,
        constant
    )

    override val maxResultShape: IntImmutableList
        get() = constant.shape

    override fun forwardPassCalculation(): TensorPointer {
        val result = executionContext.allocateForwardMemory(this, constant.shape)

        val stride =
            CommonTensorOperations.stride(constant.shape)
        System.arraycopy(constant.data, 0, result.buffer(), result.offset(), stride)

        return result
    }

    override fun leftBackwardDerivativeChainValue(): TensorPointer {
        return TrainingExecutionContext.NULL
    }

    override fun rightBackwardDerivativeChainValue(): TensorPointer {
        return TrainingExecutionContext.NULL
    }

    override val forwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(constant.shape)

    override val backwardMemoryAllocations: List<IntImmutableList>
        get() = emptyList()

    override val requiresBackwardDerivativeChainValue: Boolean
        get() = false

    override fun calculateGradientUpdate() {
        // No gradient update required
    }
}
